### PR TITLE
Feature/is 8596 adds option to register new user

### PR DESCRIPTION
**Pull request by mkrzeminskicurity**

### CODE REVIEW

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 1]
```diff
@@ -59,13 +55,21 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
...
 
-        return Optional.of(createAuthenticationResult(postRequestModel.username))
+        return Optional.of(createAuthenticationResult(postRequestModel.username, postRequestModel.register))
     }
 
-    private fun createAuthenticationResult(userName: String) = AuthenticationResult(
+    private fun createAuthenticationResult(userName: String, registrationIndicator: MutableCollection<String>?):
+            AuthenticationResult {
+        var contextAttributes = ContextAttributes.of(Attributes.of(Attribute.of("iat", Date().time)))
+        if (registrationIndicator?.isNotEmpty() == true)
```
Comment (jmdelatre) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: `== true` needless

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 2]
```diff
@@ -75,9 +79,10 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
     override fun preProcess(request: Request, response: Response): RequestModel
     {
         // set the template and model for responses on the NOT_FAILURE scope
-        response.setResponseModel(templateResponseModel(
-            singletonMap("username", userPreferencesManager.username as Any?),
-            templateName), Response.ResponseModelScope.NOT_FAILURE)
+        val data = HashMap<String, Any?>(2)
+        data["username"] = userPreferencesManager.username as Any?
```
Comment (jmdelatre) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: `as Any?` shouldn't be needed IINM

**Review by jmdelatre — APPROVED** [review]

File: src/main/resources/templates/authenticator/username/authenticate/get.vm [hunk 3]
```diff
@@ -47,7 +47,9 @@
            value="$!username">
 
     <button type="submit" class="button button-fullwidth button-primary">#message("${_templatePrefix}.view.submit")</button>
-
+    #if ($_showSignUpButton)
+        <button type="submit" class="button button-fullwidth button-primary" value="register" name="register">Register</button>
```
Comment (luisgoncalves) [inline src/main/resources/templates/authenticator/username/authenticate/get.vm]: The button's text should be resolved from a message key, right?

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/RequestModel.kt [hunk 4]
```diff
@@ -19,6 +18,7 @@ class Post(request: Request)
 {
     @NotBlank(message = "validation.error.username.required")
     val username: String = request.getFormParameterValueOrError("username")
+    var register: MutableCollection<String>? = request.getFormParameterValues("register")
```
Comment (luisgoncalves) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/RequestModel.kt]: Posting multiple values would be unexpected, right? I suggest using `getFormParameterValueOrError` and declaring the field as `String?`

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt (hunk 1 above)
Comment (luisgoncalves) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: I suggest adding a boolean getter to the request model and hiding the "string to bool" conversion in there.

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 5]
```diff
@@ -59,13 +55,21 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
...
+        return Optional.of(createAuthenticationResult(postRequestModel.username, postRequestModel.register))
     }
 
-    private fun createAuthenticationResult(userName: String) = AuthenticationResult(
+    private fun createAuthenticationResult(userName: String, registrationIndicator: MutableCollection<String>?):
+            AuthenticationResult {
+        var contextAttributes = ContextAttributes.of(Attributes.of(Attribute.of("iat", Date().time)))
+        if (registrationIndicator?.isNotEmpty() == true)
+        {
+            contextAttributes = contextAttributes.with(Attribute.of("new_account_registration", true))
```
Comment (luisgoncalves) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: Should the name of this attribute be configurable, with a default?

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 6]
```diff
@@ -35,11 +35,17 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
     private val userPreferencesManager = config.userPreferencesManager
     private val autoPostLoginHint = config.autoSubmitPreferredUserName()
     private val exceptionFactory = config.exceptionFactory
-    private val showSignUpButton = config.getShowSignUpButton()
+    private val showLinkToSetContextAttribute = config.getShowLinkToSetContextAttribute()
 
     override fun get(requestModel: RequestModel, response: Response): Optional<AuthenticationResult>
     {
-        if (autoPostLoginHint && requestModel.getRequestModel?.preferredUserName != null)
+        if (requestModel.getRequestModel?.register == true) {
```
Comment (jmdelatre) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: `== true` needless

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 7]
```diff
@@ -55,19 +61,24 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
...
-    private fun createAuthenticationResult(userName: String, registrationIndicator: MutableCollection<String>?):
+    private fun createAuthenticationResult(userName: String?, contextAttributeName: String?):
             AuthenticationResult {
         var contextAttributes = ContextAttributes.of(Attributes.of(Attribute.of("iat", Date().time)))
-        if (registrationIndicator?.isNotEmpty() == true)
+        if (contextAttributeName != null)
+        {
+            contextAttributes = contextAttributes.with(Attribute.of(contextAttributeName, true))
+        }
+        var username = userName
```
Comment (jmdelatre) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: `var username = userName ?: ""`

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 8]
```diff
@@ -35,11 +35,17 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
...
-    private val showSignUpButton = config.getShowSignUpButton()
+    private val showLinkToSetContextAttribute = config.getShowLinkToSetContextAttribute()
 
     override fun get(requestModel: RequestModel, response: Response): Optional<AuthenticationResult>
     {
-        if (autoPostLoginHint && requestModel.getRequestModel?.preferredUserName != null)
+        if (requestModel.getRequestModel?.register == true) {
+            return Optional.of(createAuthenticationResult(
+                null,
+                showLinkToSetContextAttribute.get().getContextAttributeName())
```
Comment (jmdelatre) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt:53]: `showLinkToSetContextAttribute` could be empty

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 9]
```diff
@@ -80,14 +91,18 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
...
+            data["_contextAttributeName"] = showLinkToSetContextAttribute.get().getContextAttributeName()
+        }
         response.setResponseModel(templateResponseModel(data, templateName), Response.ResponseModelScope.NOT_FAILURE)
 
         // on request validation failure, we should use the same template as for NOT_FAILURE
         response.setResponseModel(templateResponseModel(emptyMap(),
             templateName), HttpStatus.BAD_REQUEST)
 
-        return RequestModel(request, userPreferencesManager)
+        return RequestModel(request, userPreferencesManager, showLinkToSetContextAttribute.get().getContextAttributeName())
```
Comment (jmdelatre) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: `showLinkToSetContextAttribute` could be empty

File: src/main/resources/messages/sv/authenticator/username/authenticate/messages [hunk 10]
```diff
@@ -4,4 +4,5 @@ view.authenticate=Logga in
 view.username=Användarnamn
 view.submit=Nästa
 validation.error.username.required=Du måste ange ditt användarnamn
+view.register=Registrera
```
Comment (jmdelatre) [inline src/main/resources/messages/sv/authenticator/username/authenticate/messages:7]: No equivalent change in Portuguese resources?

**Review by jmdelatre — APPROVED** [review]

File: src/main/resources/messages/sv/authenticator/username/authenticate/messages (hunk 10 above)
Comment (mkrzeminskicurity) [reply src/main/resources/messages/sv/authenticator/username/authenticate/messages:7]: This plugin only seems to have en and sv

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 11]
```diff
@@ -59,25 +65,43 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
...
         // set the template and model for responses on the NOT_FAILURE scope
-        response.setResponseModel(templateResponseModel(
-            singletonMap("username", userPreferencesManager.username as Any?),
-            templateName), Response.ResponseModelScope.NOT_FAILURE)
+        val data = HashMap<String, Any?>(4)
+        data["username"] = userPreferencesManager.username
+        if (showLinkToSetContextAttribute.isPresent) {
+            data["_showLinkToSetContextAttribute"] = true
+            data["_contextAttributeMessageKey"] = showLinkToSetContextAttribute.get().getMessageKey()
+            data["_contextAttributeName"] = ADD_CONTEXT
```
Comment (luisgoncalves) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: Is `_contextAttributeName` really needed? I suggest hard-coding the parameter name in the template, as that's simpler to understand. Depending on the name of a view-model key or on a constant is exactly the same from the template's perspective.

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt [hunk 12]
```diff
@@ -40,12 +39,19 @@ class UsernameAuthenticatorRequestHandler(config: UsernameAuthenticatorPluginCon
     private val userPreferencesManager = config.userPreferencesManager
     private val autoPostLoginHint = config.autoSubmitPreferredUserName()
     private val exceptionFactory = config.exceptionFactory
+    private val showLinkToSetContextAttribute = config.getShowLinkToSetContextAttribute()
 
     override fun get(requestModel: RequestModel, response: Response): Optional<AuthenticationResult>
     {
-        if (autoPostLoginHint && requestModel.getRequestModel?.preferredUserName != null)
+        if (requestModel.getRequestModel?.register == true && showLinkToSetContextAttribute.isPresent) {
```
Comment (luisgoncalves) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/UsernameAuthenticatorRequestHandler.kt]: I suggest extracting  `requestModel.getRequestModel` to a non-nullable variable, throwing if it's not present, and use it in the different places in this method.

File: src/main/kotlin/io/curity/identityserver/plugin/username/authentication/RequestModel.kt [hunk 13]
```diff
@@ -21,7 +20,8 @@ class Post(request: Request)
     val username: String = request.getFormParameterValueOrError("username")
 }
 
-class Get(userPreferenceManager: UserPreferenceManager)
+class Get(request: Request, userPreferenceManager: UserPreferenceManager)
 {
+    var register: Boolean = request.queryParameterNames.contains(UsernameAuthenticatorRequestHandler.ADD_CONTEXT)
```
Comment (luisgoncalves) [inline src/main/kotlin/io/curity/identityserver/plugin/username/authentication/RequestModel.kt]: Detail: `register` could be renamed following up the config/parameter change.

**Review by luisgoncalves — APPROVED** [review]

**Pull request merged**